packet/https.go

```diff
@@ -42,8 +43,12 @@ func ReadTLSMessage(r io.Reader) (*TLSMessage, error) {
 		ProtoVersion: binary.BigEndian.Uint16(rawHeader[1:3]),
 		PayloadLen:   binary.BigEndian.Uint16(rawHeader[3:5]),
 	}
```

Comment:
One question about this.
```go
	header := TLSHeader{
		Type:         TLSMessageType(rawHeader[0]),
		ProtoVersion: binary.BigEndian.Uint16(rawHeader[1:3]),
		PayloadLen:   binary.BigEndian.Uint16(rawHeader[3:5]),
	}
```

According to RFC 8446, it says
```   
      enum {
          client_hello(1),
          server_hello(2),
          new_session_ticket(4),
          end_of_early_data(5),
          encrypted_extensions(8),
          certificate(11),
          certificate_request(13),
          certificate_verify(15),
          finished(20),
          key_update(24),
          message_hash(254),
          (255)
      } HandshakeType;

      struct {
          HandshakeType msg_type;    /* handshake type */
          uint24 length;             /* remaining bytes in message */
          select (Handshake.msg_type) {
              case client_hello:          ClientHello;
              case server_hello:          ServerHello;
              case end_of_early_data:     EndOfEarlyData;
              case encrypted_extensions:  EncryptedExtensions;
              case certificate_request:   CertificateRequest;
              case certificate:           Certificate;
              case certificate_verify:    CertificateVerify;
              case finished:              Finished;
              case new_session_ticket:    NewSessionTicket;
              case key_update:            KeyUpdate;
          };
      } Handshake;
```

So the first 1 byte is the type of handshake, and the following 3 bytes should be the `remaining bytes in message`

And following is the definition of ClientHello struct
```
      uint16 ProtocolVersion;
      opaque Random[32];

      uint8 CipherSuite[2];    /* Cryptographic suite selector */

      struct {
          ProtocolVersion legacy_version = 0x0303;    /* TLS v1.2 */
          Random random;
          opaque legacy_session_id<0..32>;
          CipherSuite cipher_suites<2..2^16-2>;
          opaque legacy_compression_methods<1..2^8-1>;
          Extension extensions<8..2^16-1>;
      } ClientHello;
```

If i undertood it right, shouldn't it be the 2bytes starting from `index 4` that is indicating the ProtocolVersion part like below?

```go
	header := TLSHeader{
		Type: TLSMessageType(rawHeader[0]),
		PayloadLength: TLSMessageType(rawHeader[1:4]),
		ProtoVersion: binary.BigEndian.Uint16(rawHeader[4:5]),
	}
```

Reply:
See [5.1](https://datatracker.ietf.org/doc/html/rfc8446#section-5.1).  Record Layer

Each TSL message is a record.

```
    enum {
          invalid(0),
          change_cipher_spec(20),
          alert(21),
          handshake(22),
          application_data(23),
          (255)
      } ContentType;

      struct {
          ContentType type;
          ProtocolVersion legacy_record_version;
          uint16 length;
          opaque fragment[TLSPlaintext.length];
      } TLSPlaintext;
```

Then, [4](https://datatracker.ietf.org/doc/html/rfc8446#section-4).  Handshake Protocol

 The handshake protocol is used to negotiate the security parameters
   of a connection.  Handshake messages are supplied to the TLS record
   layer, where they are encapsulated within one or more **TLSPlaintext** or
   **TLSCiphertext** structures which are processed and transmitted as
   specified by the current active connection state.

So, our `RawPayload` can be `ClientHello` or `ServerHello` or `ApplicationData`

Reply:
Yaeh makes sense. Thanks!